### ISSUE TITLE
simple: S1009 is confused by struct fields

### ISSUE THREAD

**Issue by ainar-g** [dominikh/go-tools#1136]
##  Code

```go
package main

import "fmt"

type T struct {
	nums []int
}

func firstOrNone(t T) (n int) {
	if t.nums == nil || len(t.nums) == 0 {
		return 0
	}

	return t.nums[0]
}

func main() {
	fmt.Println(firstOrNone(T{}))
}
```

##  Command

```sh
staticcheck --checks=all ./main.go
```

##  Expected

```none
main.go:6:5: should omit nil check; len() for nil slices is defined as zero (S1009)
```

## Got

No result.  But if `firstOrNone` is rewritten to accept slices, it does show the warning above.

I remember that there were several other checks that don't seem to work with struct fields that well, but I cannot find one.  Feel free to merge into that one, if it's not a figment of my imagination.

<details>
<summary>Versions</summary>

```sh
staticcheck --version
```

```none
staticcheck (devel, v0.3.0-0.dev.0.20211125235944-030b6968cbcd)
```

```sh
staticcheck --debug.version
```

```none
staticcheck (devel, v0.3.0-0.dev.0.20211125235944-030b6968cbcd)

Compiled with Go version: go1.17.2
Main module:
        honnef.co/go/tools@v0.3.0-0.dev.0.20211125235944-030b6968cbcd (sum: h1:tJGTFsNhbU6NHtXgov4i4DnUcuFnkpo9vPiXRPfiKW8=)
Dependencies:
        github.com/BurntSushi/toml@v0.4.1 (sum: h1:GaI7EiDXDRfa8VshkTj7Fym7ha+y8/XxIgD2okUIjLw=)
        golang.org/x/mod@v0.5.1 (sum: h1:OJxoQ/rynoF0dcCdI7cLPktw/hR2cueqYfjm43oqK38=)
        golang.org/x/sys@v0.0.0-20211019181941-9d821ace8654 (sum: h1:id054HUawV2/6IGm2IV8KZQjqtwAOo2CYlOToYqa0d0=)
        golang.org/x/tools@v0.1.8-0.20211111032514-c116b727ec7e (sum: h1:fsKHrXLwSIYck0izXdJysW76M60KzpYc+vKlO6s3QMY=)
        golang.org/x/xerrors@v0.0.0-20200804184101-5ec99f83aff1 (sum: h1:go1bK/D/BFZV2I8cIQd1NKEZ+0owSTG1fDTci4IqFcE=)
```

```sh
go version
```

```none
go version go1.17.2 linux/amd64
```

</details>

**Comment by dominikh** [comment]
In this particular case, we only check idents, to avoid false positives due to side effects, but we could extend this to allow certain selector expressions, too.

**Comment by spencerschrock** [comment]
This issue can be closed after d4543d1f61d555273702023457ca023fee8e3301?

```shell
staticcheck -version
staticcheck 2024.1.1 (0.5.1)
staticcheck -checks S1009 main.go
main.go:10:5: should omit nil check; len() for []int is defined as zero (S1009)
```

**Comment by dominikh** [comment]
Yes, thank you.
